carnival/carnival.ts

```diff
@@ -0,0 +1,1017 @@
+    public setTraceMulti(traceWay: tracers): void {
+
+        if (traceWay == tracers.Full) {
+            this.moon.setFlag(SpriteFlag.Invisible, true);
+            this.iter = 3;
+            this.trace = true;
+        } else if (traceWay == tracers.Part) {
+            this.moon.setFlag(SpriteFlag.Invisible, true);
+            this.iter = .3;
+            this.trace = true;
+        } else if (traceWay == tracers.Pointer) {
+            this.moon.setFlag(SpriteFlag.Invisible, true);
+            this.iter = .2;
+            this.trace = true;
+        } else if (traceWay == tracers.Cross) {
+            this.trace = false;
+            this.moon.setFlag(SpriteFlag.Invisible, false);
+        } else {
+            this.trace = false;
+            this.moon.setFlag(SpriteFlag.Invisible, false);
+        }
+    }
+
+
+    /**
+     * Set the crosshairs to distance away from center of
+     * ball in direction ball will travel
+     */
+
+    public update_crosshair(dist?: number) {
+        if (dist == undefined) { dist = 3; }
+        spriteutils.placeAngleFrom(
+            this.moon,
+            this.angle * Math.PI / -180,
+            Math.max(this.width + dist, this.height + dist),
+            this
+        )
+    }
+
+    /**
+     * Set the trace length for the estimated path in percent
+     */
+    //% weight=50
+    //% color="#b70082"
+    //% group="Ball""
+    //% blockId=set_trace_iter block="set $this trace length to $len \\%"
+    //% len.defl=50
+    //% this.defl=myBall
+    //% help=github:carnival/docs/set_trace_iter
+    public setIter(len: number): void {
+        // Make 100 percent distance = 3
+        this.iter = 3 * (len / 100);
+    }
+
+    /**
+     * Throw the throwable with the current settings
+     */
+    //% weight=70
+    //% color="#b70082"
+    //% group="Ball"
+    //% blockId=throw_it block="toss $this"
+    //% this.defl=myBall
+    //% help=github:carnival/docs/throw_it
+    public throwIt(): void {
+        this.vx = carnival.xComponent(this.angle, this.pow);
+        this.vy = carnival.yComponent(this.angle, this.pow);
+        this.ay = this.gravity;
+        this.ax = this.wind;
+    }
+
+    /**
+     * Stop the throwable at the current location
+     */
+    //% color="#b70082"
+    //% weight=70
+    //% group="Ball"
+    //% blockId=stop_it block="stop $this"
+    //% this.defl=myBall
+    //% help=github:carnival/docs/stop_it
+    public stopIt(): void {
+        this.ay = 0;
+        this.ax = 0;
+        this.vx = 0;
+        this.vy = 0;
+    }
+
+    /**
+     * Set whether to control the throwable with the arrow keys; left and right
+     * to adjust the angle, and up and down to increase / decrease power
+     * @param on whether to turn on or off this feature, eg: true
+     */
+    //% color="#b70082"
+    //% weight=80
+    //% group="Ball"
+    //% blockId=control_ball_keys block="control $this with arrow keys || $on=toggleOnOff"
```

Comment:
use `on.shadow=toggleOnOff`. same for function below

Reply:
fixed